pyclaragenomics/claragenomics/bindings/cudapoa.pyx

```diff
@@ -0,0 +1,175 @@
+                   Throws exception if error is encountered while
+                   adding POA groups.
+        """
+        if (not isinstance(poa, list)):
+            poas = [poa]
+        if (len(poa) < 1):
+            raise RuntimeError("At least one sequence must be present in POA group")
+        cdef bytes py_bytes
+        cdef char* c_string
+        cdef Group poa_group
+        cdef Entry entry
+        cdef vector[StatusType] seq_status
+        for seq in poa:
+            if (not isinstance(seq, bytes)):
+                raise RuntimeError("Sequences need to be of type 'bytes'")
+            py_bytes = seq
+            c_string = py_bytes
+            entry.seq = c_string
+            entry.weights = NULL
+            entry.length = len(seq)
+            poa_group.push_back(entry)
+        status = deref(self.batch).add_poa_group(seq_status, poa_group)
+        if status != success:
+            raise RuntimeError("Could not add POA group: " + str(status))
+        return (status, seq_status)
+
+    @property
+    def total_poas(self):
+        """
+        Get total number of POA groups added to batch.
+
+        Returns:
+            Number of POA groups added to batch.
+        """
+        return deref(self.batch).get_total_poas()
+
+    @property
+    def batch_id(self):
+        """
+        Get the batch ID of the cudapoa Batch object.
+
+        Returns:
+            Batch ID.
+        """
+        return deref(self.batch).batch_id()
+
+    def generate_poa(self):
+        """
+        Run asynchronous partial order alignment on all POA groups
+        in batch.
+        """
+        deref(self.batch).generate_poa()
+
+    def get_msa(self):
+        """
+        Get the multi-sequence alignment for each POA group.
+
+        Returns:
+            A list with MSAs for each POA group, with each MSA
+            represented as a list of alignments.
+        """
+        cdef vector[vector[string]] msa
+        cdef vector[StatusType] status
+        error = deref(self.batch).get_msa(msa, status)
+        if error == output_type_unavailable:
+            raise RuntimeError("Output type not requested during batch initialization")
+        return msa
+
+    def get_consensus(self):
+        """
+        Get the consensus for each POA group.
+
+        Returns:
+            A tuple where
+            - first element is list with consensus string for each group
+            - second element is a list of per base coverages for each consensus
+            - third element is status of consensus generation for each group
+        """
+        cdef vector[string] consensus
+        cdef vector[vector[uint16_t]] coverage
+        cdef vector[StatusType] status
+        error = deref(self.batch).get_consensus(consensus, coverage, status)
+        if error == output_type_unavailable:
+            raise RuntimeError("Output type not requested during batch initialization")
+        return (consensus, coverage, status)
```

Comment:
AFAICT consensus is a bytearray? I think we decode these elements into UTF-8 strings before returning them.

Reply:
Changed API